contrib/mpi-proxy-split/p2p_drain_send_recv.cpp

```diff
@@ -136,6 +136,14 @@ completePendingIrecvs()
         g_recvBytesByRank[worldRank] += call->count * size;
         bytesReceived += call->count * size;
         it = g_async_calls.erase(it);
+      } else {
+        // If the MPI_Test failed, we still update the iterator.
+        // Otherwise, if the message we are waiting for will be sent after
+        // the checkpoint, this can be an infinite loop.
+        // NOTE: This function will be called only if the global arrays do not
+        // match. This can happen if a second sender has sent to us, and we will
+        // receive only after the checkpoint.
+        it++;
```

Comment:
I actually thought about this solution.
Is there a chance that the peer has sent the message, but the first time `MPI_Test_internal` failed and we want to retry?
I check the implementation of `MPI_Recv`, which calls `MPI_Wait`, I guess the reason it uses a loop over there is that we don't know when the peer sent the message?

Reply:
On second thought, it doesn't matter, if that really happens, it will get back to here because of `while (!allDrained())`